### DISCUSSION TITLE
cv2.videoCapture(0) worked but cv2.videoCapture("./camera_test.avi") not work in package exe

### DISCUSSION THREAD

**Discussion by xiang-zhe** [pyinstaller/pyinstaller#6679] [Help] [answered]
win10+python38+opencv454+pyinstaller451
 
python code worked in both case:
`cfg["cameraID"] = 0` or `cfg["cameraID"] = "./camera_test.avi"`
```
cap = cv2.videoCapture(cfg["cameraID"])
ret, frame = cap.read()
```
but after package by pyinstaller, only work when `cfg["cameraID"] = 0` , **not worked** when `cfg["cameraID"] = "./camera_test.avi"`, and `os.path.isfile(cfg["cameraID"])`,return True

and opecv_vidoeio_ffmepg454_64.dll is there.
and i tried a other video "./persons.mp4"
and i tried the path like "D:\\pyinstaller\\cam_test\\persons.mp4"

error like:
```
[ERROR:0] global D:\a\opencv-python\opencv-python\opencv\modules\vide
oio\src\cap.cpp (166) cv::VideoCapture::open VIDEOIO(CV_IMAGES): raise
d OpenCV exception:

OpenCV(4.5.4-dev) D:\a\opencv-python\opencv-python\opencv\modules\vid
eoio\src\cap_images.cpp:253: error: (-5:Bad argument) CAP_IMAGES: can
't find starting number (in the name of file): ./camera_test.avi in function '
cv::icvExtractPattern'

> failed to open video capture >>: ./camera_test.avi _(this is my print)_
Traceback (most recent call last):
  File "test_cam.py", line 10, in <module>
TypeError: __init__() should return None, not 'int' _(this is my return 1 when not cap.isOpened() )_
[5984] Failed to execute script 'test_cam' due to unhandled exception!
```

test code like:
```
jsrw = JsonRW()
cfg = jsrw.read_json("./config.json")
print(cfg)
print(cfg["camera"]["cam_id"])
print("file",os.path.isfile(cfg["camera"]["cam_id"]))
source = CamerReader2(camera_id=cfg["camera"]["cam_id"], 
            height=cfg["camera"]["cam_height"], 
            width=cfg["camera"]["cam_width"], 
            framerate=cfg["camera"]["cam_framerate"])
show = 1
for frame in source:
    if show:
        cv2.namedWindow("im")
        width, height = 1920, 1080
        cv2.resizeWindow("im", width, height)
        cv2.imshow("im",cv2.cvtColor(frame, cv2.COLOR_BGR2RGB))
        cv2.waitKey(1)
cv2.destroyAllWindows()

class CamerReader2(IterableDataset):
    def __init__(self, camera_id=0, height=1080, width=1920, framerate=30):
        if isinstance(camera_id, int):
            self.cap = cv2_VideoCapture(camera_id, cv2_CAP_DSHOW)
            # use motion-jpeg codec to a fourcc code
            #self.cap.set(cv2_CAP_PROP_FOURCC, cv2_VideoWriter_fourcc('M', 'J', 'P', 'G'))
            # use 4K@60
            self.cap.set(cv2_CAP_PROP_FPS, framerate)
            self.cap.set(cv2_CAP_PROP_FRAME_WIDTH, width)
            self.cap.set(cv2_CAP_PROP_FRAME_HEIGHT, height)
        else:
            self.cap = cv2_VideoCapture(camera_id)
        if not self.cap.isOpened():
            print(f'> failed to open video capture >>: {camera_id}')
            return 1
    def __iter__(self):
        while self.cap.isOpened():
            ret, frame = self.cap.read()
            if ret:
                yield  cv2_cvtColor(frame, cv2_COLOR_BGR2RGB)
                #yield frame[:,:,::-1]
            else:
                break
```

**Comment by rokm** [comment, marked as answer]
Hmm, judging by

```
[ERROR:0] global D:\a\opencv-python\opencv-python\opencv\modules\videoio\src\cap.cpp (166) cv::VideoCapture::open VIDEOIO(CV_IMAGES): raised OpenCV exception:

OpenCV(4.5.4-dev) D:\a\opencv-python\opencv-python\opencv\modules\videoio\src\cap_images.cpp:253: error: (-5:Bad argument) CAP_IMAGES: can't find starting number (in the name of file): ./camera_test.avi in function 'cv::icvExtractPattern'
```

OpenCV is for some reason treating the given file as image filename pattern instead of the video file (i.e., trying to use CV_IMAGES backend instead of ffmpeg one).

----

Can you setting `OPENCV_VIDEOIO_DEBUG=1` environment variable to see more related debug messages and see what's going on?

Or, you can also try listing all available video stream backends at the start of your program, to see if the ffmpeg one is properly recognized:
<details><summary>Example</summary><p>

```python
import cv2

backends = cv2.videoio_registry.getStreamBackends()
print("Video stream backends:", backends)
for backend in backends:
    print(f"Backend {backend}")
    print(f" - name: {cv2.videoio_registry.getBackendName(backend)}")
    built_in = cv2.videoio_registry.isBackendBuiltIn(backend)
    print(f" - built-in: {built_in}")
    if not built_in:
        try:
            print(f" - plugin version: {cv2.videoio_registry.getStreamBackendPluginVersion(backend)}")
        except cv2.error:
            print(f" - plugin version: failed to query")
```

</p></details>

**Reply by xiang-zhe** [reply]
Hi， @rokm ，thank you for the reply， i run the example，
result from python code， like：
```
Video stream backends: (1900, 1800, 2300, 2000, 2200)
Backend 1900
 - name: FFMPEG
 - built-in: False
 - plugin version: ('FFmpeg OpenCV Video I/O Capture plugin', 1, 1)
Backend 1800
 - name: GSTREAMER
 - built-in: False
 - plugin version: failed to query
Backend 2300
 - name: INTEL_MFX
 - built-in: False
 - plugin version: failed to query
Backend 2000
 - name: CV_IMAGES
 - built-in: True
Backend 2200
 - name: CV_MJPEG
 - built-in: True
```
result from exe， like：
```
Video stream backends: (1900, 1800, 2300, 2000, 2200)
Backend 1900
 - name: FFMPEG
 - built-in: False
 - plugin version: failed to query
Backend 1800
 - name: GSTREAMER
 - built-in: False
 - plugin version: failed to query
Backend 2300
 - name: INTEL_MFX
 - built-in: False
 - plugin version: failed to query
Backend 2000
 - name: CV_IMAGES
 - built-in: True
Backend 2200
 - name: CV_MJPEG
 - built-in: True
```
it seems like only the difference is the plugin version in Backend 1900，but i didn't know what‘s means，whether it's matter.

**Reply by rokm** [reply]
I think it does matter. In my test example, the exe also returns:

```
Video backends: (1900, 1800, 2300, 1400, 2000, 2200)
Backend 1900
 - name: FFMPEG
 - built-in: False
 - plugin version: ('FFmpeg OpenCV Video I/O Capture plugin', 1, 1)
Backend 1800
 - name: GSTREAMER
 - built-in: False
 - plugin version: failed to query
Backend 2300
 - name: INTEL_MFX
 - built-in: False
 - plugin version: failed to query
Backend 1400
 - name: MSMF
 - built-in: True
Backend 2000
 - name: CV_IMAGES
 - built-in: True
Backend 2200
 - name: CV_MJPEG
 - built-in: True
```

It does so even if I move the collected `opencv_videoio_ffmpeg454_64.dll` from `cv2` sub-directory to top-level directory, but if I remove it completely, I get the same output as you ("failed to query" for ffmpeg plugin version). So I think in your case, the FFMPEG plugin is either not properly collected (although you did mention that it is there?), or it is not recognized by OpenCV for some reason.

**Reply by rokm** [reply]
How did you install OpenCV, though? `pip install opencv-python` or some other method? I find it a bit odd that you don't have backend 1400/MSMF while I do...

**Reply by xiang-zhe** [reply]
it’s interesting that i also move `opencv_videoio_ffmpeg454_64.dll` from `cv2` sub-directory to top-level directory, then it works，but still have not 1400， and also
```
Backend 1900
 - name: FFMPEG
 - built-in: False
 - plugin version: ('FFmpeg OpenCV Video I/O Capture plugin', 1, 1)
```
and about a month ago， after package exe，the exe will error with  can't found opencv and some others（i don‘t know why， maybe the env broken that day）；after that i need `add --paths cv2 --add-binary onnxruntime.dll` when pyinstallering，whether that's matter i don't know，or maybe i need add `--add-binary opencv_videoio_ffmpeg454_64.dll`;
i installed opencv with pip，in fact when removing `opencv_videoio_ffmpeg454_64.dll`，the exe still can read camera，so the
cv2.videocapture() not call ffmepg;
so why it can't found when in `cv2` folder , but it works both camera and video now when i moved `opencv_videoio_ffmpeg454_64.dll`  from `cv2` sub-directory to top-level directory，thank you very much.

**Reply by rokm** [reply]
I think if you use `--paths cv2`, then the `opencv_videoio_ffmpeg454_64.dll` needs to be in top-level directory instead of `cv2` directory. Because in that case, `cv2` will not be a package, but rather a module (extension) only.

`--paths cv2` might have been needed as a work-around for OpenCV 4.5.4.58, but should not be needed for later versions, thanks to pyinstaller/pyinstaller-hooks-contrib#354.

**Reply by xiang-zhe** [reply]
because of the error` OpenCV loader: missing configuration file: ['config.py']`, according to issues, it need add either `--paths cv2` or downgraded to `opencv==4.5.3.56`; so i added `--paths cv2`, but according to 

> I think if you use `--paths cv2`, then the `opencv_videoio_ffmpeg454_64.dll` needs to be in top-level directory instead of `cv2` directory. Because in that case, `cv2` will not be a package, but rather a module (extension) only.

so i need move the the `opencv_videoio_ffmpeg454_64.dll` to top-level directory by myself(but it won't work when a file package exe); now i downgraded the opencv to 4.5.3.56 and it works well (also without `--paths cv2`) for both camera and video, thank you again.
